### PR TITLE
feat(jsx) Add support for non-react JSX transformer #92

### PR DESCRIPTION
**Pull request by Cmdv**
Unsure if this what you had in mind @staltz.
Created a constructor function but I've not written tests wanted to make sure this is the right way to go first.

@calebmer If you could also take a look if you think I've done this right?

### CODE REVIEW

File: jsx.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+var jsx = require('babel-plugin-transform-jsx');
+var h = require("./src/virtual-hyperscript");
+
+/**
+ * Constructor function for https://www.npmjs.com/package/babel-plugin-transform-jsx
+ * Docs are in ./src/cycle-dom
+ */
+module.exports = function (attrs) {
+   var convertedJSX = jsx(attrs);
```
Comment (calebmer) [inline jsx.js]: You don't need to require the module. You do the necessary transforms on the JSX object here. So for example:

```js
function (jsxObject) {
  return h(
    jsxObject.elementName,
    jsxObject.attributes,
    jsxObject.children
  )
}
```

Comment (Cmdv) [inline jsx.js]: I knew I had things confused!! thank you @calebmer I'll make the amends :+1:

File: jsx.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+var jsx = require('babel-plugin-transform-jsx');
```
Comment (calebmer) [inline jsx.js]: Just remember this isn't needed 😊

**Comment by staltz** [comment]
Yes @Cmdv looks correct :)

File: jsx.js (hunk 2 above)
Comment (Cmdv) [inline jsx.js]: damn it! took it out the `package.json` missed it in this file, will amend :)

**Comment by calebmer** [comment]
Looks great to me 😊👍

**Comment by Cmdv** [comment]
Cool sorted @calebmer :smiley: 

@staltz not entirely sure I can write a test for this, unless you can think of something?

File: docs/api.md [hunk 3]
```diff
@@ -83,6 +84,22 @@ attributes key like `<tag attributes={{'data-custom-attr': 'foo'}} />`.
...
+{
+ "plugins": [["babel-plugin-transform-jsx", { "module": "@cycle/dom/jsx" }]]
+}
+```
+Note there is now no need to add `\/\* @jsx *\/\` at the top of the files using JSX.
+
+For more information read the documentations [here]((https://www.npmjs.com/package/babel-plugin-transform-jsx)).
+
+- - -
+
```
Comment (staltz) [inline docs/api.md]: @Cmdv did you write to this markdown document directly or did you generate it using `npm run docs`? We can't do the former.

**Comment by staltz** [comment]
@Cmdv check in tests/browser/render.js there is a test for JSX using hJSX. It's enough to make a test file that imports this jsx file, creates some JSX of some divs, then inspects what is the output. The npm `pretest-browser` script uses `babelify` to transpile the tests, so maybe you can point to a test babel configuration that will then specify the plugins.

File: docs/api.md (hunk 3 above)
Comment (Cmdv) [inline docs/api.md]: No it's inside `src/cycle-dom.js` but when I put it into `jsx.js` I had to adjust the `make-api-docs.js` to add `jsx.js` into an array but the output from running the script was a whole new section with header titles and links, I'll investigate if that file can just append to the current titles etc.

**Comment by Cmdv** [comment]
No ta problem I'll make some test this evening :)

File: docs/api.md (hunk 3 above)
Comment (staltz) [inline docs/api.md]: Ah I see

**Comment by calebmer** [comment]
Ok, I'm currently working to help on tests, and there's a complication. Cycle DOM is using babel `5.x` whereas `babel-plugin-transform-jsx` depends on babel `6.x`. I'm going to make a test which downloads a custom version of babel in a subdirectory of `test`. The upgrade path wouldn't be hard, just keep the `.babelrc` file I'm going to add and remove the `package.json` specifying the babel `6.x` dependency. Tell me if you have a preference for another solution.

I'll ping this discussion again when I have a PR for @Cmdv's branch.

**Comment by Cmdv** [comment]
@calebmer cool this sounds familiar as to where I was falling over when I was trying to get the tests going! thanks for the helping hand though it's appreciated :+1:

**Comment by calebmer** [comment]
Opened a PR to @Cmdv's branch with passing tests, this can be found [here](https://github.com/Cmdv/cycle-dom/pull/2/files). I think we are on the 3rd level of git right now. Do we need to go deeper? 😊

If you ever upgrade to babel

**Comment by Cmdv** [comment]
@calebmer not think that will be adequate, thank you very much for the help on those :smiley:

**Comment by Cmdv** [comment]
@calebmer think the test are struggling with finding the right path to the transform-jsx: `Error: Cannot find module './fixtures/transform-jsx' from '/home/travis/build/cyclejs/dom/test/browser`

I'll investigate later :+1:

**Comment by calebmer** [comment]
It's because the `test-ci` script doesn't build the file like the `test-browser` script.

**Comment by Cmdv** [comment]
@calebmer do you think making a separate script file just for this test would be a way around it?

**Comment by calebmer** [comment]
As long as the script gets called by both `test-ci` and `test-browser`, yeah. Look at my commit's diff to see what the script would need to contain.

**Comment by staltz** [comment]
Why is `test/browser/fixtures/transform-jsx/index.js` in .gitignore? Isn't that the source of the problem?

**Comment by calebmer** [comment]
The `index.js` file is built at test time from `index.source.jsx` using a newer version of Babel.

**Comment by calebmer** [comment]
On second thought, if you don't care about testing the actual file build (which is my responsibility anyway) you could just test that the contract works. So a test written like this:

```js
jsx({
  elementName: 'h1',
  attributes: { className: 'foo' },
  children: ['Hello, world!']
})
```

Where `jsx` is your function then trust my package to successfully build to this format. Just another option if the install/build step feels too much.

**Comment by staltz** [comment]
Hi, thanks for your contributions :bow: 

Check the Travis build, though, there are some errors:

>  `Cannot find module './fixtures/transform-jsx' from '/home/travis/build/cyclejs/dom/test/browser'`

I suppose it's because you are doing

```js
const app = require('./fixtures/transform-jsx');
```

while it should be

```js
const app = require('./fixtures/transform-jsx/index');
```

**Comment by calebmer** [comment]
@staltz I think the more likely reason is that the `index.js` file isn't building in the CI environment. I added a build step to `test-browser` for building the JSX file with Babel 6 and this build step is not run in the `test-ci` script.

I think this decision is best deferred to yourself or @Cmdv for deciding where the JSX file build step should go in the test process.

**Comment by Cmdv** [comment]
cheers @calebmer I don't mind making the amends see if that works, if you decide where you'd like the file to live @staltz

**Comment by staltz** [comment]
> file isn't building in the CI environment. I added a build step to test-browser for building the JSX file with Babel 6 and this build step is not run in the test-ci script.

Yeah, it's not, and it should. Sounds simple to solve

**Pull request closed without merging**